### ISSUE TITLE
reduced_decision_function_trainer2 does not actually reduce number of basis vectors.

### ISSUE THREAD

**Issue by Andyalevy** [davisking/dlib#1195]
## Expected Behavior
When using the reduced_decision_function_trainer2 with a svm_c_trainer, with a set num_bv, I'd expect that if the trained decision function already has fewer basis vectors than the num_bv, it would not attempt to reduce it, but just return a decision function the same way the svm_c_trainer would have.

## Current Behavior
Regardless of the resulting basis vectors of the non-reduced decision function, the reduced_decision_function_trainer2 is iterating over the trainer until it achieves the received num_bv.

## Steps to Reproduce
 - Declare any type of trainer (I used svm_c_trainer).
 - Train it.
 - Check the number of basis vectors it has.
 - Declare a reduced_decision_function_trainer2 with your trainer and a num_bv bigger than the returned number of basis vectors.
 - Train it.
 - Compare the number of basis vectors in the decision function trained by the reduced_decision_function_trainer2 with the number of basis vectors of the decision function trained by the original trainer. 

* **Version**:  19.7
* **Where did you get dlib**: dlib.net
* **Platform**:  Windows 64-bit
* **Compiler**:  Visual Studio 2015

**Comment by davisking** [comment]
Yeah that's probably not the best behavior.  You should submit a PR :)
